contents/java/Java.md

```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+
+### The Basics of Reflections
+
+Before getting started with reflections in Java, it is important to realise that classes itself is an object. Every
+unique `Object` is assigned an immutable `Class` object by the JVM. This is fundamentally different from *instances* of
+a class. The class object itself holds information about its name, the package it resides in and other important
+information while an instance of a class holds the instanced values and methods as defined in the class.
+
+Take for example the following class:
+
+```java
+public class Student {
+
+	private final String name;
+	private final String gender;
+
+	public Student(String name, String gender) {
+		this.name = name;
+		this.gender = gender;
+	}
+
+	public String getName() {
+		return name;
+	}
+
+	public String getGender() {
+		return gender;
+	}
+}
+```
+
+An instance of a class can be created a usual using the new keyword
+
+```java
+Student john = new Student("John Doe", "Male");
+```
+
+However, it is also possible to get information about the class itself because the class itself is an Object like so:
+
+```java
+Class<Student> studentClass = Student.class;
+```
+
+It also means that you can store any Class object in any data structure for future retrieval. This is the main entry
+point for Java's reflections. You can now get the name of the class, create new instances of the class, observe its
+public/private fields - the possibilities are endless!
+
+### Getting Started
+
+There are many webpages dedicated to explaining the details of reflections in Java; so this will not repeat what is
+being made readily available on the web. One good place to start is [http://www.javaworld.com/article/2077015/java-se/take-an-in-depth-look-at-the-java-reflection-api.html](http://www.javaworld.com/article/2077015/java-se/take-an-in-depth-look-at-the-java-reflection-api.html)
+
+What this page will do however is to give some examples to demonstrate what reflections can do and in then give 
+you ideas on how you can use reflections in your project.
+
+One important point is that while Java reflections are powerful, its implementations are anything but. There are however
+some libraries out there to make life easier such as the [Google's Guava
+library](https://github.com/google/guava/wiki/ReflectionExplained) which contains many utility methods and classes to
+make your life easier.
+
+#### Applications
+
+You might have learnt from your Software Engineering module that the Observer pattern can be used for Objects that are
+interested to get notified if the state of another object is changed. However, a you don't want to create a troublesome
+bidirectional dependency between two unrelated objects that have no business talking to each other. This is when the
+Observer pattern becomes useful.
+
+One prime example of the implementation of the `Observer` pattern is the Google Events bus used in `AddressBook Level 4`.
```

Comment:
It is probably better not to assume students have taken CS2103 specifically (The book is aimed at typical SE students, but not necessarily SE students in SoC. -- from the 3281 schedule page) and maybe just add a reference to the actual code with a link?